cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ApplicationVerificationStage.java

```diff
@@ -211,27 +211,32 @@ private void verifyWorkflowNode(ApplicationSpecification appSpec, WorkflowSpecif
   }
 
   private void verifyWorkflowFork(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
-                                  WorkflowNode node) {
+                                  WorkflowNode node, Set<String> existingNodeNames) {
     WorkflowForkNode forkNode = (WorkflowForkNode) node;
     Preconditions.checkNotNull(forkNode.getBranches(), String.format("Fork is added in the Workflow '%s' without" +
                                                                        " any branches", workflowSpec.getName()));
 
     for (List<WorkflowNode> branch : forkNode.getBranches()) {
-      verifyWorkflowBranch(appSpec, workflowSpec, branch);
+      verifyWorkflowNodeList(appSpec, workflowSpec, branch, existingNodeNames);
     }
   }
 
   private void verifyWorkflowCondition(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
-                                       WorkflowNode node) {
+                                       WorkflowNode node, Set<String> existingNodeNames) {
     WorkflowConditionNode condition = (WorkflowConditionNode) node;
-    verifyWorkflowBranch(appSpec, workflowSpec, condition.getIfBranch());
-    verifyWorkflowBranch(appSpec, workflowSpec, condition.getElseBranch());
+    verifyWorkflowNodeList(appSpec, workflowSpec, condition.getIfBranch(), existingNodeNames);
+    verifyWorkflowNodeList(appSpec, workflowSpec, condition.getElseBranch(), existingNodeNames);
   }
 
-  private void verifyWorkflowBranch(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
-                                    List<WorkflowNode> branch) {
-    for (WorkflowNode n : branch) {
-      verifyWorkflowNode(appSpec, workflowSpec, n);
+  private void verifyWorkflowNodeList(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
+                                      List<WorkflowNode> nodeList, Set<String> existingNodeNames) {
+    for (WorkflowNode n : nodeList) {
+      if (existingNodeNames.contains(n.getNodeId())) {
+        throw new RuntimeException(String.format("Node with the name '%s' added multiple times in workflow '%s'.",
```

Comment:
nit: to be consistent, should we just say ``Node '%s' already exists in workflow '%s'``?